src/Lecture2.hs

```diff
@@ -264,7 +345,30 @@ data EvalError
 It returns either a successful evaluation result or an error.
 -}
 eval :: Variables -> Expr -> Either EvalError Int
-eval = error "TODO"
+eval _ (Lit value) = Right value
+eval variables (Var variable) = if value == Nothing
+                                then Left (VariableNotFound variable)
+                                else Right (fromJust value)
+  where value = lookup variable variables
+        fromJust               :: Maybe a -> a
+        fromJust (Just a) =  a
+        fromJust Nothing  =  error "fromJust: Nothing"
```

Comment:
You're using the partial function `fromJust` here and you compare your value with `Nothing` first. It's correct, however, such an approach is considered an antipattern. You can use `case-of` instead to write the correct-by-construction function:

```suggestion
eval variables (Var variable) = case lookup variable variables of
    Nothing -> Left (VariableNotFound variable)
    Just x  -> Right x
```